language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToFunctionCodeAction.java

```diff
@@ -0,0 +1,593 @@
+         * 2. The number of local variables in (1.) is less or equal to 1.
+         *
+         * */
+        boolean isRangeExtractable = updatingButNotDeclaredLocVarSymbolsInRange.size() == 0
+                && locVarSymbolsDeclaredOrUpdatingInRangeAndReferredAfterRange.size() <= 1;
+
+        // here we decide whether the selected range is extractable by the usages of symbols
+        if (!isRangeExtractable) {
+            return Optional.empty();
+        }
+
+        Optional<VariableSymbol> updatingVar = Optional.empty();
+        if (locVarSymbolsDeclaredOrUpdatingInRangeAndReferredAfterRange.size() == 1) {
+            updatingVar = locVarSymbolsDeclaredOrUpdatingInRangeAndReferredAfterRange.stream().findFirst();
+        }
+        return Optional.of(updatingVar);
+    }
+
+    private Range getRangeAfterHighlightedRange(CodeActionContext context, Node enclosingNode) {
+        Position endPosOfEnclosingNode = PositionUtil.toPosition(enclosingNode.lineRange().endLine());
+        Position endPosOfStatements
+                = new Position(endPosOfEnclosingNode.getLine(), endPosOfEnclosingNode.getCharacter() - 1);
+        return new Range(context.range().getEnd(), endPosOfStatements);
+    }
+
+    private List<Symbol> getVarAndParamSymbolsWithinRangeForStmts(CodeActionContext context, Node enclosingNode) {
+        List<Symbol> argsSymbolsForExtractFunction = new ArrayList<>();
+        List<Symbol> visibleSymbolsToStartPosOfRange = getVisibleSymbols(context, context.range().getStart());
+
+        List<Symbol> localVarSymbolsDeclarationsBeforeStartOfRange = visibleSymbolsToStartPosOfRange.stream()
+                .filter(symbol -> symbol.kind() == SymbolKind.VARIABLE && symbol.getLocation().isPresent())
+                // this check is done to filter the local variables defined inside the function
+                .filter(symbol -> PositionUtil
+                        .isWithinLineRange(symbol.getLocation().get().lineRange(), enclosingNode.lineRange()))
+                .collect(Collectors.toList());
+
+        HashSet<Symbol> paramAndLocVarSymbolsVisibleToRange = visibleSymbolsToStartPosOfRange.stream()
+                .filter(symbol -> symbol.kind() == SymbolKind.PARAMETER).collect(Collectors.toCollection(HashSet::new));
+        paramAndLocVarSymbolsVisibleToRange.addAll(localVarSymbolsDeclarationsBeforeStartOfRange);
+
+        paramAndLocVarSymbolsVisibleToRange.forEach(symbol -> {
+            Optional<Location> anyLocationReferred = context.currentSemanticModel().get().references(symbol).stream()
+                    // check whether the symbols are referred inside the selected range
+                    .filter(location -> PositionUtil.isRangeWithinRange(PositionUtil
+                            .getRangeFromLineRange(location.lineRange()), context.range()))
+                    .findAny();
+            if (anyLocationReferred.isPresent()) {
+                argsSymbolsForExtractFunction.add(symbol);
+            }
```

Comment:
I am not adding references to the list, it's the symbol, hence we cannot do this

Reply:
reference is the symbol. Let's change the variable name to symbol then.

Reply:
Initially I tried to do this without the foreach but couldn't figure out how to include this `argsSymbolsList.add(symbol)` part to the original stream(since we cannot do `argsSymbolsForExtractFunction.add(symbol))` inside the `ifPrfesent()`). Can you please suggest the whole lambda expression? (in your code line, the `reference` is a location not the symbol we want to add to the argsList)